src/index.ts

```diff
@@ -108,54 +116,102 @@ function createPreprocessor(
+	let start = 0;
+	bundlerMap.on("start", ev => {
+		if (singleBundle) {
+			start = ev.time;
+			log.info(`Compiling to ${ev.file}...`);
+		} else if (!pendingBundles.size) {
+			start = ev.time;
+			log.info(`Compiling...`);
+		}
+		pendingBundles.add(ev.file);
+	});
+
+	bundlerMap.on("stop", ev => {
+		pendingBundles.delete(ev.file);
+	});
+
+	bundlerMap.on("done", ev => {
+		pendingBundles.delete(ev.file);
+		if (singleBundle || pendingBundles.size === 0) {
+			log.info(`Compiling done (${formatTime(ev.endTime - start)})`);
+		}
 	});
 
 	const buildBundle = debounce(() => {
 		// Prevent service closed message when we are still processing
 		if (stopped) return;
 		testEntryPoint.write();
+
+		const bundle = bundlerMap.get(testEntryPoint.file);
 		return bundle.write();
 	}, bundleDelay);
 
 	return async function preprocess(content, file, done) {
 		// Karma likes to turn a win32 path (C:\foo\bar) into a posix-like path (C:/foo/bar).
 		// Normally this wouldn't be so bad, but `bundle.file` is a true win32 path, and we
 		// need to test equality.
-		const filePath = path.normalize(file.originalPath);
+		let filePath = path.normalize(file.originalPath);
 
-		testEntryPoint.addFile(filePath);
+		if (singleBundle) {
+			testEntryPoint.addFile(filePath);
+			filePath = testEntryPoint.file;
+		} else {
+			bundlerMap.addPotential(filePath);
+		}
+
+		const bundle = bundlerMap.get(filePath);
 		bundle.dirty();
-		await buildBundle();
 
-		// Turn the file into a `dom` type with empty contents to get Karma to
-		// inject the contents as HTML text. Since the contents are empty, it
-		// effectively drops the script from being included into the Karma runner.
-		file.type = "dom";
-		done(null, "");
+		if (singleBundle) {
+			await buildBundle();
+			// Turn the file into a `dom` type with empty contents to get Karma to
+			// inject the contents as HTML text. Since the contents are empty, it
+			// effectively drops the script from being included into the Karma runner.
+			file.type = "dom";
+			done(null, "");
+		} else {
+			const res = await bundlerMap.read(filePath);
+
+			(file as any).sourceMap = res.map;
+			done(null, res.code);
+		}
 	};
 }
 createPreprocessor.$inject = [
 	"config",
 	"emitter",
-	"karmaEsbuildLogger",
 	"karmaEsbuildEntryPoint",
-	"karmaEsbuildBundler",
+	"karmaEsbuildBundlerMap",
+	"logger",
 ];
 
-function createMiddleware(bundle: Bundle) {
+function createMiddleware(bundlerMap: BundlerMap, config: karma.ConfigOptions) {
 	return async function (
 		req: IncomingMessage,
 		res: ServerResponse,
 		next: () => void,
 	) {
-		const match = /^\/absolute([^?#]*?)(\.map)?(?:\?|#|$)/.exec(req.url || "");
-		if (!match) return next();
+		const match = /^\/(?:absolute|base\/)([^?#]*?)(\.map)?(?:\?|#|$)/.exec(
+			req.url || "",
+		);
+		if (!match) {
+			return next();
+		}
 
-		const filePath = path.normalize(match[1]);
-		if (filePath !== bundle.file) return next();
+		const absolutePath = path.isAbsolute(match[1])
```

Comment:
I believe this could just be done by by checking to see if the url started with `/base/`? We could turn `(?:absolute|base\/)` into `(absolute|base\/)` and then check the first match, too.

But why is this necessary? Shouldn't all paths be `/absolute`?